Add map function (#4806)
brimdata/zed: docs/language/functions/map.md

### Function

> &emsp; **map** &mdash; calls a function on each element of an array and returns the results

Reviewer: ```suggestion
&emsp; **map** &mdash; apply a function to each element of an array or set
```

### Synopsis

```

> map(v: array|set, f: function) -> array|set

Reviewer 2: @nwt: Can you say more about why this is your preference? It feels strange to me for the data type to be changed as a side effect of an operation like this. If we end up going with your suggestion, then I'd be in favor of the docs mentioning your `|[...map(v,f)]|` workaround in the same doc right below this function signature. Those of us who work on Zed knowing something is possible is not the same thing as it being something average users will think to put together. I think of the spread operator as a more advanced part of the language, so leaving this as an "exercise for the reader" doesn't sit with me right.

> map(v: array|set, f: function) -> array|set

Author: Yeah I agree with @philrz. The map function is really targeting each element in the array/set and it feels weird to have it change the type of the container.

> map(v: array|set, f: function) -> array|set

Reviewer: I suggest this because having `map(some_large_set, is_interesting)` return `|[true, false]|` isn't useful.

> map(v: array|set, f: function) -> array|set

Reviewer 2: @nwt: Maybe I'm missing some context, but I'm not sure how that example is the one that would make the case for this. I admittedly don't find myself reaching for the `set` data type very often, but to reach for our old friend Zeek that's always happened to have `set` data, I can see places to use this, e.g. with the [zed-sample-data](https://github.com/brimdata/zed-sample-data):

```
$ zq -version
Version: v1.10.0-15-g7be4063c

$ zq -Z 'lower_analyzers:=map(analyzers, lower) | cut analyzers,lower_analyzers | head 1' files.log.gz
{
    analyzers: |[
        "MD5",
        "SHA1"
    ]|,
    lower_analyzers: [
        "md5",
        "sha1"
    ]
}
```

Indeed, that shows it's turning the set into an array even in the current state of the branch, and I'm still not sure why that's helpful here to me as a user.

> map(v: array|set, f: function) -> array|set

Author: In the case of `map(some_large_set, is_interesting)` it seems like the thing to do would be to convert `some_large_set` into an array if you don't want the output values to get squashed. The normal case IMHO would be something like what @philrz is proposing.

```

### Description

> The _map_ function applies function f to every element in array/set v and
> returns the augmented array/set. Function f must be a function that takes

Reviewer: Nits: "or" is classier than "/". And "augmented" suggests that values returned by f are added to v.
```suggestion
The _map_ function applies function f to every element in array or set v and
returns an array or set of the results. Function f must be a function that takes
```

only one argument. A user defined function can be used for f.

### Examples

Upper case each element of an array:

```mdtest-command
echo '["foo","bar","baz"]' | zq -z 'yield map(this, upper)' -
```
=>
```mdtest-output
["FOO","BAR","BAZ"]
```

Using a user defined function to convert an epoch float to a time:

```mdtest-input udf.zed
func floatToTime(x): ( cast(x*1000000000, <time>) )
yield map(this, floatToTime)
```

```mdtest-command
echo '[1697151533.41415,1697151540.716529]' | zq -z -I udf.zed -

> ```

Reviewer: Nit: I think this would be clearer without udf.zed.
```mdtest-command
echo '[1697151533.41415,1697151540.716529]' | zq -z '
  func floatToTime(x): ( cast(x*1000000000, <time>) )
  yield map(this, floatToTime)
' -
```
